test/run_e2e_test.py

```diff
@@ -0,0 +1,195 @@
+
+    :param args:
+    :return:
+    """
+    expected_nodes = ['/listener', '/talker'] + EXPECTED_LIFECYCLE_NODES
+
+    rclpy.init(args=args)
+    node = rclpy.create_node('check_for_expected_nodes_test')
+    actual_nodes = node.get_node_names_and_namespaces()
+    node.destroy_node()
+    rclpy.shutdown()
+
+    expected_count = len(expected_nodes)
+    actual_count = 0
+    for item in actual_nodes:
+        found_node = '/' + item[0]
+        if found_node in expected_nodes:
+            actual_count += 1
+            expected_nodes.remove(found_node)
+
+    if expected_count != actual_count:
+        raise Exception('Failed to enumerate expected nodes. Found #' + actual_count + ': ', actual_nodes)
+    else:
+        print_test('check_for_expected_nodes success')
+
+
+def check_lifecycle_node_enumeration() -> None:
+    """
+    Check that all lifecycle nodes exist.
+
+    This requires the ros2lifecycle dependency.
+    :return:
+    """
+    stream = os.popen('ros2 lifecycle nodes')
+    output = stream.read().splitlines()
+    if output != EXPECTED_LIFECYCLE_NODES:
+        raise Exception('check_lifecycle_node_enumeration failed: ' + str(output))
+    else:
+        print_test('check_lifecycle_node_enumeration success')
+
+
+def check_lifecycle_node_state() -> None:
+    """
+    Check that each lifecycle node is active.
+
+    This requires the ros2lifecycle dependency.
+    :return:
+    """
+    for lifecycle_node in EXPECTED_LIFECYCLE_NODES:
+        stream = os.popen('ros2 lifecycle get ' + str(lifecycle_node))
+        output = stream.read().rstrip()
+
+        if output != EXPECTED_LIFECYCLE_STATE:
+            raise Exception('check_lifecycle_node_state:'
+                            + lifecycle_node +
+                            ' not in expected state: '
+                            + str(output))
+
+    print_test('check_lifecycle_node_state success')
+
+
+def check_for_statistic_publications(args=None) -> None:
+    """
+    Check that all nodes publish a statistics message.
+
+    :param args:
+    :return:
+    """
+    future = Future()
+    rclpy.init(args=args)
+    node = StatisticsListener(future, EXPECTED_LIFECYCLE_NODES)
+    rclpy.spin_until_future_complete(node, future, timeout_sec=TIMEOUT_SECONDS)
+    node.destroy_node()
+    rclpy.shutdown()
+
+    if not node.received_all_published_stats:
+        raise Exception('check_for_statistic_publications failed. Absent publisher(s): '
+                        + str(node.expected_lifecycle_nodes))
+    else:
+        print_test('check_for_statistic_publications success')
+
+
+def main(args=None) -> int:
+    """
+    Run all the e2e tests. This exits on the first failure encountered.
+
+    :param args:
+    :return: 0 if all tests pass, 1 if any fail
+    """
+    retval = 1
+    split_command = LAUNCH_COMMAND.split()
+    print_test('Executing: ' + str(split_command))
+    process = subprocess.Popen(split_command)
+    time.sleep(2)
```

Comment:
Not quite. Right now I'm trying to figure out an independent status / log entry to signify the launch process has proceeded enough (with a timeout in case it didn't).

Reply:
A better approach is to reuse the node enumeration test with a max attempt counter. If the nodes appear before the count is exceeded then we can proceed with the checks, otherwise fast fail.